File: src/cleansia_android/customer-app/src/main/java/cz/cleansia/customer/core/booking/CashEligibility.kt
package cz.cleansia.customer.core.booking

/**
 * Whether a booking may be paid in cash — `BookingPolicy.AllowsCash`: a signed-in customer, on a booking
 * the server says one cleaner does alone. The crew is the quote's `requiredEmployees` for the selection on
 * screen, never a client estimate; null means no quote describes that selection yet. The server re-decides
 * on create, so this only decides what the payment step offers.
 */
sealed interface CashEligibility {
    data object Available : CashEligibility

    data object NeedsAccount : CashEligibility

    data class NeedsCard(val requiredCleaners: Int) : CashEligibility

    data object Pending : CashEligibility

    /** A verdict that takes cash away. An unknown crew is not one — the choice is kept until it is known. */
    val isRefused: Boolean get() = this is NeedsAccount || this is NeedsCard

    companion object {
        fun resolve(signedIn: Boolean, requiredEmployees: Int?): CashEligibility = when {
            requiredEmployees != null && requiredEmployees > 1 -> NeedsCard(requiredEmployees)
            !signedIn -> NeedsAccount
            requiredEmployees == 1 -> Available
            else -> Pending
        }
    }
}

File: src/cleansia_ios/CleansiaCustomer/Sources/Features/Booking/CashEligibility.swift
import Foundation

/// `BookingPolicy.AllowsCash`. `requiredEmployees` is the quote's crew for the selection on screen —
/// never assigned crew or spare seats — and nil while no quote describes that selection. The server
/// re-decides on create, so this only decides what the payment step offers.
enum CashEligibility: Equatable {
    case available
    case needsAccount
    case needsCard(requiredCleaners: Int)
    case pending

    static let refusalCode = "order.cash_not_available"

    static func resolve(signedIn: Bool, requiredEmployees: Int?) -> CashEligibility {
        if let requiredEmployees, requiredEmployees > 1 {
            return .needsCard(requiredCleaners: requiredEmployees)
        }
        guard signedIn else { return .needsAccount }
        return requiredEmployees == 1 ? .available : .pending
    }

    /// An unknown crew is not a refusal: a cash choice is kept until the crew is known.
    var refusesCash: Bool {
        switch self {
        case .needsAccount, .needsCard: true
        case .available, .pending: false
        }
    }
}
